truffle/src/main/ruby/core/rubinius/api/shims/thread.rb

```diff
@@ -15,4 +42,43 @@ def recursive_objects
     @recursive_objects ||= {}
   end
 
+  def raise_prim(exc)
+    Rubinius.primitive :thread_raise
+    Kernel.raise PrimitiveFailure, "Thread#raise primitive failed"
+  end
+
+  private :raise_prim
+
+  def raise(exc=undefined, msg=nil, trace=nil)
+    return self unless alive?
+
+    if undefined.equal? exc
+      no_argument = true
+      exc         = nil
+    end
+
+    if exc.respond_to? :exception
+      exc = exc.exception msg
+      Kernel.raise TypeError, 'exception class/object expected' unless Exception === exc
+      exc.set_backtrace trace if trace
+    elsif no_argument
+      exc = RuntimeError.exception nil
+    elsif exc.kind_of? String
+      exc = RuntimeError.exception exc
+    else
+      Kernel.raise TypeError, 'exception class/object expected'
+    end
+
+    if $DEBUG
+      STDERR.puts "Exception: #{exc.message} (#{exc.class})"
+    end
+
+    if self == Thread.current
+      Kernel.raise exc
+    else
+      exc.capture_backtrace! 2 unless exc.backtrace?
+      raise_prim exc
+    end
+  end
+
```

Comment:
Small correction `Thread#raise` is defined in `bootstrap/thread` we do not have it in our repo though, which was to reason why I've putted into existing `shims/thread`. You are right that it's error prone (if we later copy the `bootsrap`) and it would be better to have it loaded in extra file. Moving.